File: packages/serverless/lib/plugins/aws/offline/lib/runners/wrappers/ruby/invoke.rb
require 'json'

# Fake AWS Lambda context for Ruby handlers. Mirrors the AWS Lambda Ruby
# runtime's context interface — handlers read context.function_name,
# context.aws_request_id, context.get_remaining_time_in_millis, etc.
class FakeLambdaContext
  attr_reader :aws_request_id, :client_context, :function_name,
              :function_version, :identity, :invoked_function_arn,
              :log_group_name, :log_stream_name, :memory_limit_in_mb

  def initialize(context:)
    @aws_request_id = context['awsRequestId']
    @client_context = context['clientContext']
    @function_name = context['functionName']
    @function_version = context['functionVersion']
    @identity = context['identity']
    @invoked_function_arn = context['invokedFunctionArn']
    @log_group_name = context['logGroupName']
    @log_stream_name = context['logStreamName']
    @memory_limit_in_mb = context['memoryLimitInMB']
    @timeout = context['timeout']

    @created_time = Time.now
  end

  def get_remaining_time_in_millis
    [@timeout * 1000 - ((Time.now - @created_time) * 1000).round, 0].max
  end
end

if __FILE__ == $PROGRAM_NAME
  unless ARGV[0] && ARGV[1]
    warn 'Usage: invoke.rb <handler_path> <handler_name>'
    exit 1
  end

  handler_path = ARGV[0]
  handler_name = ARGV[1]

  # Load the handler module once. `require` is relative to the current
  # working directory; the Node-side runner spawns us with cwd set to the
  # handler's directory and passes the basename, mirroring the M5b Python
  # cwd+basename pattern.
  require("./#{handler_path}")

  # Handler name is either a global method or a Module::Class.method form.
  # Resolve once outside the loop.
  handler_method, handler_class = handler_name.split('.').reverse
  handler_class ||= 'Kernel'
  resolved_class = Object.const_get(handler_class)

  # NOTE: the community plugin calls `attach_tty` here to allow `binding.pry`
  # in user code. We deliberately omit it — our streaming child owns stdin
  # for the JSON-line protocol; reopening fd 0 to /dev/tty would either fail
  # or starve subsequent reads.

  # Streaming protocol: one JSON event per line in, one JSON envelope per
  # result line out. Mirrors the M5b Python wrapper's structure so the
  # long-lived child can serve multiple invocations on the same functionKey.
  loop do
    line = $stdin.gets
    break if line.nil? # parent closed stdin → exit gracefully

    input = JSON.parse(line)
    context = FakeLambdaContext.new(context: input['context'])
    result = resolved_class.send(handler_method,
                                 event: input['event'],
                                 context: context)

    data = {
      # Identifier to distinguish the result envelope from puts/log output.
      '__offline_payload__' => result
    }

    $stdout.puts(data.to_json)
    $stdout.flush
  end
end

File: packages/serverless/lib/plugins/aws/offline/lib/runners/wrappers/python/invoke.py
import base64
import subprocess
import argparse
import json
import logging
import sys
import os
from time import strftime, time
from importlib import import_module


class FakeLambdaContext(object):
    def __init__(self, name='Fake', version='LATEST', timeout=6, **kwargs):
        self.name = name
        self.version = version
        self.created = time()
        self.timeout = timeout
        for key, value in kwargs.items():
            setattr(self, key, value)

    def get_remaining_time_in_millis(self):
        return int(max((self.timeout * 1000) - (int(round(time() * 1000)) - int(round(self.created * 1000))), 0))

    @property
    def function_name(self):
        return self.name

    @property
    def function_version(self):
        return self.version

    @property
    def invoked_function_arn(self):
        return getattr(self, 'invokedFunctionArn', None) \
            or 'arn:aws:lambda:us-east-1:000000000000:function:' + self.name

    @property
    def memory_limit_in_mb(self):
        return getattr(self, 'memoryLimitInMB', None) or '1024'

    @property
    def aws_request_id(self):
        return getattr(self, 'awsRequestId', None) or '1234567890'

    @property
    def log_group_name(self):
        return getattr(self, 'logGroupName', None) \
            or '/aws/lambda/' + self.name

    @property
    def log_stream_name(self):
        return getattr(self, 'logStreamName', None) \
            or strftime('%Y/%m/%d') + '/[$' + self.version + ']58419525dade4d17a495dceeeed44708'

    @property
    def log(self):
        return sys.stdout.write


logging.basicConfig()

parser = argparse.ArgumentParser(
    prog='invoke',
    description='Runs a Lambda entry point (handler) with an optional event',
)

parser.add_argument('handler_path',
                    help=('Path to the module containing the handler function,'
                          ' omitting ".py". IE: "path/to/module"'))

parser.add_argument('handler_name', help='Name of the handler function')

if __name__ == '__main__':
    args = parser.parse_args()

    # this is needed because you need to import from where you've executed sls
    sys.path.append('.')

    module = import_module(args.handler_path.replace(os.sep, '.'))
    handler = getattr(module, args.handler_name)

    # Keep a reference to the original stdin so that we can continue to receive
    # input from the parent process.
    stdin = sys.stdin

    if sys.platform != 'win32':
        try:
            if sys.platform != 'darwin':
                subprocess.check_call('tty', stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        except (OSError, subprocess.CalledProcessError):
            pass
        else:
            # Replace stdin with a TTY to enable pdb usage. /dev/tty is only
            # available when the process has a controlling terminal; when
            # spawned headless (CI, Jest, detached child) the open() raises
            # OSError. Failing here is harmless — `stdin` (the original) is
            # already captured above, and the protocol loop reads from it.
            try:
                sys.stdin = open('/dev/tty')
            except OSError:
                pass

    while True:
        input = json.loads(stdin.readline())

        context = FakeLambdaContext(**input.get('context', {}))
        result = handler(input['event'], context)

        data = {
            # identifier to distinguish the result envelope from print()/log output
            '__offline_payload__': result
        }

        if hasattr(result, 'body') and isinstance(result['body'], bytes):
            data['__offline_payload__']['body'] = base64.b64encode(result['body']).decode('utf-8')
            data['isBase64Encoded'] = True

        sys.stdout.write(json.dumps(data))
        sys.stdout.write('\n')
        sys.stdout.flush()
